onnx/src/ops/ml/mod.rs

```diff
@@ -0,0 +1,330 @@
+            .or_else(|err| node.bail(&format!("{}", err)))?;
+        trees.push(tree);
+    }
+
+    // build the tree ensemble
+
+    let base_scores = base_values.as_ref().map(Vec::as_slice);
+    let ensemble = TreeEnsemble::build(&trees, aggregate_fn, post_transform, base_scores)
+        .or_else(|err| node.bail(&format!("{}", err)))?;
+
+    Ok(ensemble)
+}
+
+pub fn register_all_ops(reg: &mut OpRegister) {
+    reg.insert("TreeEnsembleClassifier", TreeEnsembleClassifier::parse);
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum ClassLabels {
+    Ints(Vec<i64>),
+    Strings(Vec<String>),
+}
+
+#[derive(Debug, Clone)]
+pub struct TreeEnsembleClassifier {
+    ensemble: TreeEnsemble,
+    class_labels: ClassLabels,
+}
+
+impl TreeEnsembleClassifier {
+    fn parse(node: &NodeProto) -> TractResult<Box<Op>> {
+        let ensemble = parse_tree_ensemble(node, true)?;
+        let class_labels = match node.get_attr_opt_slice::<i64>("classlabels_int64s")? {
+            Some(int_labels) => ClassLabels::Ints(int_labels.into()),
+            _ => {
+                let str_labels = node.get_attr_opt_vec::<String>("classlabels_strings")?;
+                ClassLabels::Strings(str_labels.unwrap())
+            }
+        };
+        Ok(Box::new(Self { ensemble, class_labels }))
+    }
+}
+
+impl Op for TreeEnsembleClassifier {
+    fn name(&self) -> Cow<str> {
+        "TreeEnsembleClassifier".into()
+    }
+
+    fn noutputs(&self) -> usize {
+        2
+    }
+}
+
+impl StatelessOp for TreeEnsembleClassifier {
+    fn eval(&self, mut inputs: TVec<SharedTensor>) -> TractResult<TVec<SharedTensor>> {
+        Err("WIP".into())
+    }
+}
+
+impl InferenceRulesOp for TreeEnsembleClassifier {
+    fn rules<'r, 'p: 'r, 's: 'r>(
+        &'s self, s: &mut Solver<'r>, inputs: &'p [TensorProxy], outputs: &'p [TensorProxy],
+    ) -> InferenceResult {
+        check_input_arity(&inputs, 1)?;
+        check_output_arity(&outputs, 2)?;
+
+        s.equals(&inputs[0].rank, 2)?;
+        s.equals(&outputs[0].rank, 1)?;
+        s.equals(&outputs[0].rank, 1)?;
+
+        s.given_3(
+            &inputs[0].shape,
+            &outputs[0].shape,
+            &outputs[1].shape,
+            move |s, s_in0, s_out0, s_out1| {
+                s.equals(s_in0[0], s_out0[0])?;
+                s.equals(s_out0[0], s_out1[0])?;
+                if let Ok(k) = s_in0[1].to_integer() {
+                    self.ensemble.check_n_features(k as _)?;
+                }
+                if let Ok(k) = s_out1[1].to_integer() {
+                    s.equals(k, self.ensemble.n_classes() as i32)?;
+                }
+                Ok(())
+            },
+        )?;
+
+        s.given(&inputs[0].datum_type, move |_, dt| {
+            Ok(match dt {
+                DatumType::F32 | DatumType::F64 | DatumType::I64 | DatumType::I32 => (),
+                _ => bail!("invalid input type for tree ensemble classifier: {:?}", dt),
+            })
+        })?;
```

Comment:
This is actually better than most rules() I wrote. I think we need a "one_of" rule and a "castable_to" rule for datum types.

Reply:
Yea, I was looking for `one_of()` rule and couldn't find it. Would be sure nice to have it in the solver.

As for `castable_to()`, it probably would be less useful if we want to stick strictly to onnx spec, since sometimes it's more like "what onnx thinks it's castable to" than "what it really is castable to".

Reply:
Well, on top of that, there are cases where i need to diverge from the onnx spec in terms in typing: in the case of voice inputs, we often need to reason with a "time" dimension which is either a variable of the input, or an actual (infinite) stream. You may have noticed the TDim stuff... So in these situation, I need onnx (or tf) default integer types (i64 for onnx, i32 for tf) to be upcasted to a TDim that can compute expression like "S/2+4" and the like...